src/net/rtsp/sender.cpp

```diff
@@ -0,0 +1,143 @@
+#include "sender.hpp"
+
+
+namespace shar::net::rtsp {
+
+ResponseSender::ResponseSender(Context context, IpAddress ip, Port port)
+  : Context(std::move(context))
+  , m_ip(ip)
+  , m_port(port)
+  , m_context()
+  , m_current_socket(m_context)
+  , m_acceptor(m_context)
+{}
+
+void ResponseSender::run(Receiver<Unit> packets) {
+  tcp::Endpoint endpoint{ m_ip, m_port };
+  m_acceptor.open(endpoint.protocol());
+  m_acceptor.set_option(tcp::Acceptor::reuse_address(true));
+  m_acceptor.bind(endpoint);
+  m_acceptor.listen(10);
+  start_accepting();
+
+  while (!m_running.expired() ) {
+    while(auto unit = packets.try_receive()){}
+    m_context.run_for(Milliseconds(50));
+  }
+}
+
+void ResponseSender::shutdown() {
+  m_running.cancel();
+}
+
+void ResponseSender::start_accepting(){
+  m_acceptor.async_accept(m_current_socket, [this](const ErrorCode& ec) {
+    if (ec) {
+      m_logger.error("Acceptor error: {}", ec.message());
+      if (m_clients.empty()){
+        shutdown();
+      }
+      return;
+    }
+    auto client_id = static_cast<std::size_t>(m_current_socket.native_handle());
+    m_logger.info("Client {} connected.", client_id);
+
+    auto [client_pos, is_emplaced] = m_clients.emplace(client_id,  std::move(m_current_socket));
+    if (is_emplaced) {
+      receive_request(client_pos);
+    }
+
+    m_current_socket = tcp::Socket(m_context);
+
+    start_accepting();
+    });
+}
+
+
+ResponseSender::Client::Client(tcp::Socket&& socket)
+  : m_socket(std::move(socket))
+  , m_buffer(4096, 0)
+  , m_received_bytes(0)
+  , m_sent_bytes(0)
+  , m_headers(10)
+{
+}
+
+void ResponseSender::disconnect_client(ClientPos client_pos) {
+  auto& client = client_pos->second;
+  client.m_socket.shutdown(tcp::Socket::shutdown_both);
+  client.m_socket.close();
+  m_clients.erase(client_pos);
+}
+
+void ResponseSender::receive_request(ClientPos client_pos) {
+  auto& client = client_pos->second;
+  auto id = client_pos->first;
+  if (client.m_received_bytes == 4096) {
+    m_logger.info("Client {}: buffer overflow", id);
+    disconnect_client(client_pos);
+    return;
+  }
+  auto buffer = span(client.m_buffer.data() + client.m_received_bytes, 4096 - client.m_received_bytes);
+  client.m_socket.async_receive(buffer, [this, &client, id](const ErrorCode& ec, const std::size_t size) mutable {
+    auto client_pos = m_clients.find(id);
+
+    if (client_pos == m_clients.end()) {
+      m_logger.info("Client {} not found", id);
+      return;
+    }
+    if (ec) {
+      m_logger.info("Socket read error (Client {})", id);
```

Comment:
It should be error instead of info, also it would be nice to have description of error here